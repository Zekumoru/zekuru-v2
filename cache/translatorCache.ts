import { AuthorizationError, Translator, TranslatorOptions } from 'deepl-node';
import { Collection } from 'discord.js';
import GuildKey, { IGuildKey } from '../models/GuildKey';
import { loadLanguages } from '../translation/languages';
import { errorDebug } from '../utils/logger';
import { decrypt, encrypt } from '../utils/crypt';

export const translatorOptions: TranslatorOptions = {
  appInfo: {
    appName: 'Zekuru-v2 Demo',
    appVersion: '1.0.0',
  },
  minTimeout: 500, // 500 ms
  maxRetries: 10,
};

const cacheTranslators = new Collection<string, Translator>();

const fetchOrCreateGuildKey = async (guildId: string, hashedApiKey: string) => {
  // fetch from db, if not exists, create new
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (guildKey) return [guildKey, false] as const;

  // save api key to db
  const newGuildKey = new GuildKey({
    id: guildId,
    key: hashedApiKey,
  });
  await newGuildKey.save();

  return [newGuildKey, true] as const;
};

const set = async (guildId: string, apiKey: string) => {
  // create translator client
  const translator = new Translator(apiKey, translatorOptions);

  // this should throw error if invalid api key
  await translator.getUsage();
  await loadLanguages(translator);

  // encrypt api key
  const encryptedApiKey = encrypt(apiKey);

  // update in db
  const [guildKey, isNewlyCreated] = await fetchOrCreateGuildKey(
    guildId,
    encryptedApiKey
  );

  if (!isNewlyCreated) {
    guildKey.overwrite({
      id: guildId,
      key: encryptedApiKey,
      createdAt: guildKey.createdAt,
    } as IGuildKey);
    await guildKey.save();
  }

  // update cache
  cacheTranslators.set(guildId, translator);
};

const get = async (guildId: string) => {
  // get from cache
  const translatorCache = cacheTranslators.get(guildId);
  if (translatorCache) return translatorCache;

  // if not exists in cache, fetch from api key from db
  // then create new translator and set to cache
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (!guildKey) return;

  try {
    const apiKey = decrypt(guildKey.key);
    const translator = new Translator(apiKey, translatorOptions);

    // load languages
    await loadLanguages(translator);

    // update cache
    cacheTranslators.set(guildId, translator);
    return translator;
  } catch (error) {
    // delete invalid api key from db and "sign out" user by removing it from cache
    if (error instanceof AuthorizationError) {
      await unset(guildId);
    }
    errorDebug(error);
  }
};

const unset = async (guildId: string) => {
  // check if it exists first
  if ((await get(guildId)) == null) return;

  // delete from db
  await GuildKey.deleteOne({ id: guildId });

  // remove from cache
  cacheTranslators.delete(guildId);
};

export default { set, unset, get };
