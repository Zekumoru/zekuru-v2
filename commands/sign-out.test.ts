import signOut from './sign-out';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import translatorCache from '../cache/translatorCache';
import createMockInteractionResponse from '../test/createMockInteractionResponse';

jest.mock('../cache/translatorCache');

describe('/sign-out command', () => {
  const mockTranslatorCache = jest.mocked(translatorCache);

  afterEach(() => {
    jest.resetAllMocks();
  });

  it(`should notify user that this command is only for servers`, async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await signOut.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user that they're already signed out`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue(null as any);

    await signOut.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should not sign out if cancelled`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({} as any);
    const mockUpdate = jest.fn();
    interaction.reply.mockResolvedValue(
      createMockInteractionResponse(
        {},
        { customId: 'cancel', update: mockUpdate }
      )
    );

    await signOut.execute(interaction);

    expect(mockTranslatorCache.unset).not.toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(mockUpdate.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should not sign out if timed out`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({} as any);
    interaction.reply.mockResolvedValue({
      awaitMessageComponent: jest.fn().mockImplementation(() => {
        throw new Error('Generic error!');
      }),
    } as any);

    await signOut.execute(interaction);

    expect(mockTranslatorCache.unset).not.toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(interaction.editReply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should sign out`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({} as any);
    const mockUpdate = jest.fn();
    interaction.reply.mockResolvedValue(
      createMockInteractionResponse(
        {},
        { customId: 'confirm', update: mockUpdate }
      )
    );

    await signOut.execute(interaction);

    expect(mockTranslatorCache.unset).toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(mockUpdate.mock.calls[0][0]).toMatchSnapshot();
  });
});
