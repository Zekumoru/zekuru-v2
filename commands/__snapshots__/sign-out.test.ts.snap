// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/sign-out command should not sign out if cancelled 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "cancel",
          "emoji": undefined,
          "label": "No",
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "confirm",
          "emoji": undefined,
          "label": "Yes, sign out",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "Are you sure to sign out the bot? This will stop translating any further messages.",
}
`;

exports[`/sign-out command should not sign out if cancelled 2`] = `
{
  "components": [],
  "content": "Signing out has been cancelled.",
}
`;

exports[`/sign-out command should not sign out if timed out 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "cancel",
          "emoji": undefined,
          "label": "No",
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "confirm",
          "emoji": undefined,
          "label": "Yes, sign out",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "Are you sure to sign out the bot? This will stop translating any further messages.",
}
`;

exports[`/sign-out command should not sign out if timed out 2`] = `
{
  "components": [],
  "content": "Confirmation not received within 30 seconds, cancelling.",
}
`;

exports[`/sign-out command should notify user that they're already signed out 1`] = `
{
  "content": "You are already signed out.",
}
`;

exports[`/sign-out command should notify user that this command is only for servers 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/sign-out command should sign out 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "cancel",
          "emoji": undefined,
          "label": "No",
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "confirm",
          "emoji": undefined,
          "label": "Yes, sign out",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "Are you sure to sign out the bot? This will stop translating any further messages.",
}
`;

exports[`/sign-out command should sign out 2`] = `
{
  "components": [],
  "content": "You have successfully signed out the bot.",
}
`;
